lib/SILOptimizer/Analysis/CallerAnalysis.cpp

```diff
@@ -10,65 +10,309 @@
+    }
+#endif
+  }
+}
+
+#ifndef NDEBUG
+static void checkCallSiteInvariants(
+    SmallSetVector<SILInstruction *, 8> &callSitesThatMustBeVisited,
+    SILInstruction &i,
+    const SmallPtrSetImpl<SILInstruction *> &visitedCallSites) {
+  if (auto apply = FullApplySite::isa(&i)) {
+    if (apply.getCalleeFunction() && !visitedCallSites.count(&i)) {
+      callSitesThatMustBeVisited.insert(&i);
+    }
+    return;
+  }
+
+  // Make sure that we are in sync with looking for partial apply callees.
+  if (auto *pai = dyn_cast<PartialApplyInst>(&i)) {
+    if (pai->getCalleeFunction() && !visitedCallSites.count(&i)) {
+      callSitesThatMustBeVisited.insert(pai);
+    }
+    return;
+  }
+}
+#endif
+
+void CallerAnalysis::processFunctionCallSites(SILFunction *callerFn) {
+  // First grab our caller info so that we can store back references
+  // from our callerFn to the calleeFn so that we can invalidate all
+  // callee info about our caller efficiently.
+  FunctionInfo &callerInfo = unsafeGetFunctionInfo(callerFn);
 
-        // Update the callee information for this function.
-        FunctionInfo &CallerInfo = FuncInfos[F];
-        CallerInfo.Callees.insert(CalleeFn);
-        
-        // Update the callsite information for the callee.
-        FunctionInfo &CalleeInfo = FuncInfos[CalleeFn];
-        CalleeInfo.Callers.insert(F);
+#ifndef NDEBUG
+  llvm::SmallPtrSet<SILInstruction *, 8> visitedCallSites;
+  llvm::SmallSetVector<SILInstruction *, 8> callSitesThatMustBeVisited;
+#endif
+
+  // Scan the caller function and search for full or partial apply
+  // sites in the caller function.
+  for (auto &block : *callerFn) {
+    for (auto &i : block) {
+#ifndef NDEBUG
+      // If this is a call site that we visited as part of seeing a different
+      // function_ref, skip it. We know that it has been processed correctly.
+      if (visitedCallSites.count(&i))
         continue;
-      }
-      if (auto *PAI = dyn_cast<PartialApplyInst>(&II)) {
-        SILFunction *CalleeFn = PAI->getCalleeFunction();
-        if (!CalleeFn)
-          continue;
+#endif
 
-        // Update the callee information for this function.
-        FunctionInfo &CallerInfo = FuncInfos[F];
-        CallerInfo.Callees.insert(CalleeFn);
-        
-        // Update the partial-apply information for the callee.
-        FunctionInfo &CalleeInfo = FuncInfos[CalleeFn];
-        int &minAppliedArgs = CalleeInfo.PartialAppliers[F];
-        int numArgs = (int)PAI->getNumArguments();
-        if (minAppliedArgs == 0 || numArgs < minAppliedArgs) {
-          minAppliedArgs = numArgs;
+      // Try to find the apply sites for this specific FRI.
+      if (auto *fri = dyn_cast<FunctionRefInst>(&i)) {
+        // If we found any interesting information, update our internal state.
+        if (auto result = findLocalApplySites(fri)) {
+          findApplySites(callSitesThatMustBeVisited, callerFn, callerInfo, fri,
```

Comment:
I don't think this code will build in Release mode. It looks like you might want to define a class/struct to hold the state required to process a caller function. That would solve the NDEBUG problem and address the `findApplySite` code smell where you're passing in a bunch of context that isn't specific to this invocation.